### PR TITLE
feat(graphqlmetrics): implement more efficient batching of clickhouse entries

### PR DESCRIPTION
**Pull request by Noroth**
- **fix: implement batching behavior for the metrics service**
- **chore: add environment config for batch processor**
- **chore: use worker pool for processing**
- **chore: improve batch processor logging**

## Motivation and Context

Clickhouse works better with bigger batches. Currently for each request that is sent to the metrics endpoints, the provided data would be sent to Clickhouse directly. 

Instead of doing that, the requests should be collected for a certain amount of time and then sent as a bigger chunk.

### CODE REVIEW

File: graphqlmetrics/config/config.go [hunk 1]
```diff
@@ -35,6 +35,10 @@ type Config struct {
 	IsPrometheusEnabled  bool   `default:"false" envconfig:"PROMETHEUS_ENABLED"`
 	PrometheusListenAddr string `default:"127.0.0.1:8088" envconfig:"PROMETHEUS_LISTEN_ADDR"`
 	PrometheusPath       string `default:"/metrics" envconfig:"PROMETHEUS_PATH"`
+
+	MaxCostThreshold        int           `default:"10000" envconfig:"MAX_COST_THRESHOLD"`
```
Comment (StarpTech) [inline graphqlmetrics/config/config.go:39]: Please prefix the environment variables to understand that they belong together.

File: graphqlmetrics/cmd/main.go [hunk 2]
```diff
@@ -149,7 +156,7 @@ func main() {
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		ms.Shutdown(cfg.ShutdownDelay)
```
Comment (StarpTech) [inline graphqlmetrics/cmd/main.go:152]: Why did you remove it?

File: graphqlmetrics/pkg/batch/processor.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+// receiveElement receives a batch of buffer adds them to the buffer.
+// If the buffer is full, the batch processing will be invoked.
+func (p *Processor[T]) receiveElement(element T) {
+	p.buffer = append(p.buffer, element)
+
+	if p.costFunc != nil {
+		p.currentBufferCost += p.costFunc(element)
+	}
+
+	p.logger.Debug("current buffer cost", zap.Int("currentCost", p.currentBufferCost))
```
Comment (StarpTech) [inline graphqlmetrics/pkg/batch/processor.go:171]: Please use snake_case everywhere.

File: graphqlmetrics/core/metrics_service.go [hunk 4]
```diff
@@ -33,355 +43,377 @@ type MetricsService struct {
...
+		)
+	})
+
+	return nil
 }
 
-func (s *MetricsService) Shutdown(deadline time.Duration) {
-	s.pool.StopAndWaitFor(deadline)
+// calculateRequestCost the total number of entries of metrics batch.
+func calculateRequestCost(item SchemaUsageRequestItem) int {
```
Comment (StarpTech) [inline graphqlmetrics/core/metrics_service.go:411]: Please add a unit test for this.

File: graphqlmetrics/pkg/batch/processor.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+		p.receiveElement(batch)
+	}
+
+	p.logger.Info("processing remaining items in the buffer", zap.Int("buffer_size", len(p.buffer)))
+	if len(p.buffer) > 0 {
+		p.process()
+	}
+}
+
+func (p *Processor[T]) canAcceptBatch() bool {
```
Comment (StarpTech) [inline graphqlmetrics/pkg/batch/processor.go:140]: How does it determine that the processor is ready to accept when it always return true?

File: graphqlmetrics/pkg/batch/processor.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+		case <-ticker.C:
+			p.logger.Debug("interval completed - processing batch")
+			p.process()
+		case element := <-p.queue:
+			p.logger.Debug("receiving element from queue",
+				zap.Any("element", element),
+				zap.Int("queue_size", len(p.queue)),
+			)
+
+			p.receiveElement(element)
```
Comment (StarpTech) [inline graphqlmetrics/pkg/batch/processor.go:109]: You did not reset the timer after we have sent a batch.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: graphqlmetrics/pkg/batch/processor.go (hunk 3 above)
Comment (Noroth) [reply graphqlmetrics/pkg/batch/processor.go:171]: the metrics_server wasn't using `snake_case` that's why I followed the existing naming. I will update everything

File: graphqlmetrics/cmd/main.go (hunk 2 above)
Comment (Noroth) [reply graphqlmetrics/cmd/main.go:152]: Initially I removed the pool and wanted to ensure to process everything in the batch processor. It does make sense to have a timeout here. I will re-add it.

File: graphqlmetrics/pkg/batch/processor.go (hunk 3 above)
Comment (StarpTech) [reply graphqlmetrics/pkg/batch/processor.go:171]: Thank you

File: graphqlmetrics/pkg/batch/processor.go (hunk 5 above)
Comment (StarpTech) [reply graphqlmetrics/pkg/batch/processor.go:140]: I got it. It used to block new requests on shutdown.

**Review by StarpTech — APPROVED** [review]
LGTM
